mm2src/mm2_main/tests/docker_tests/swap_watcher_tests.rs

```diff
@@ -0,0 +1,965 @@
+use crate::docker_tests::docker_tests_common::{eth_distributor, generate_erc20_coin,
+                                               generate_eth_coin_with_random_privkey};
+use crate::integration_tests_common::*;
+use crate::{generate_utxo_coin_with_privkey, generate_utxo_coin_with_random_privkey, random_secp256k1_secret,
+            SecretKey};
+use coins::coin_errors::ValidatePaymentError;
+use coins::utxo::UtxoCommonOps;
+use coins::{FoundSwapTxSpend, MarketCoinOps, MmCoinEnum, SearchForSwapTxSpendInput, SendTakerPaymentArgs,
+            SendWatcherRefundsPaymentArgs, SwapOps, WatcherOps, WatcherValidateTakerFeeInput,
+            EARLY_CONFIRMATION_ERR_LOG, INVALID_RECEIVER_ERR_LOG, INVALID_SENDER_ERR_LOG, OLD_TRANSACTION_ERR_LOG};
+use common::{block_on, now_ms, DEX_FEE_ADDR_RAW_PUBKEY};
+use futures01::Future;
+use mm2_main::mm2::lp_swap::{dex_fee_amount_from_taker_coin, MAKER_PAYMENT_SENT_LOG, MAKER_PAYMENT_SPEND_FOUND_LOG,
+                             MAKER_PAYMENT_SPEND_SENT_LOG, TAKER_PAYMENT_REFUND_SENT_LOG, WATCHER_MESSAGE_SENT_LOG};
+use mm2_number::MmNumber;
+use mm2_test_helpers::for_tests::{enable_eth_coin, eth_jst_conf, eth_testnet_conf, mm_dump, mycoin1_conf, mycoin_conf,
+                                  start_swaps, MarketMakerIt, Mm2TestConf, ETH_SEPOLIA_NODE,
+                                  ETH_SEPOLIA_SWAP_CONTRACT, ETH_SEPOLIA_TOKEN_CONTRACT};
+use num_traits::Pow;
+use serde_json::Value as Json;
+use std::thread;
+use std::time::Duration;
+use uuid::Uuid;
+
+fn round(num: f64, digits: u8) -> f64 {
+    let digits: f64 = (10.).pow(digits);
+    (digits * num).round() / digits
+}
+
+fn enable_eth_and_jst(mm_node: &MarketMakerIt) {
+    dbg!(block_on(enable_eth_coin(
+        mm_node,
+        "ETH",
+        ETH_SEPOLIA_NODE,
+        ETH_SEPOLIA_SWAP_CONTRACT,
+        Some(ETH_SEPOLIA_SWAP_CONTRACT)
+    )));
+
+    dbg!(block_on(enable_eth_coin(
+        mm_node,
+        "JST",
+        ETH_SEPOLIA_NODE,
+        ETH_SEPOLIA_SWAP_CONTRACT,
+        Some(ETH_SEPOLIA_SWAP_CONTRACT)
+    )));
+}
+
+fn get_balance(mm_node: &MarketMakerIt, ticker: &str) -> String {
+    let rc = block_on(mm_node.rpc(&json!({
+    "userpass": mm_node.userpass,
+    "method": "my_balance",
+    "coin": ticker
+    })))
+    .unwrap();
+    assert!(rc.0.is_success(), "!my_balance: {}", rc.1);
+
+    let json: Json = serde_json::from_str(&rc.1).unwrap();
+    json["balance"].as_str().unwrap().to_string()
+}
+
+fn get_balance_f64(mm_node: &MarketMakerIt, ticker: &str) -> f64 {
+    get_balance(mm_node, ticker).parse::<f64>().unwrap()
+}
+
+#[test]
+#[ignore]
```

Comment:
Could you please clarify why this and other tests are ignored?

Reply:
There were some problems that I left for later since it already took too much time. The problems are about the tests themselves, not about the functions being tested. The tests are based on the comparison of alice/bob balances before/after the watcher interruption. One problem is that due to the ethereum gas fee, it's not possible to know the expected balance after alice/bob sends transactions. I tried to overcome this by rounding the least significant digits, which worked most of the time while failing occasionally. This was the case when I ran the tests individually, but when I used cargo test, they all failed and I couldn't find out the reason yet. And even though this balance comparison problem is solved, we still need to constantly fill the alice/bob/watcher with ethereum due to gas fee spending and for this reason it might be the best to keep these tests ignored and use only when necessary.

Reply:
Ok, thanks for clarification!